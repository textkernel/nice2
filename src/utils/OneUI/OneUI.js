import cssVarsPonyfill from 'css-vars-ponyfill';

const isInternetExplorer11 = () => {
    const ua = window.navigator.userAgent;
    return ua.indexOf('Trident/') > 0;
};

const DEFAULT_LOADING_TIMEOUT = 2000;

class ThemeLoader {
    /**
     * Loads the theme and the CSS vars ponyfill if necessary
     *
     * @param {Object} themeConfig - Object that contains the configuration to initialize the theme
     * @param {string} themeConfig.themeURL - URL that serves the theme file
     * @param {number} themeConfig.maxTime - Max time to wait for the theme to be loaded
     * @param {Object} themeConfig.ponyfillOptions - Set of options that can be used to configure the ponyfill. Options: https://www.npmjs.com/package/css-vars-ponyfill#options
     */
    static init({ themeURL = '', maxTime = DEFAULT_LOADING_TIMEOUT, ponyfillOptions } = {}) {
        const loadTheme = Promise.all([
            ThemeLoader.applyTheme(themeURL),
            ThemeLoader.startCssVarsPonyfill(ponyfillOptions)
        ]);

        const timeout = new Promise((resolve, reject) =>
            setTimeout(
                () => reject(new Error(`Theme "${themeURL}" not loaded. Loading time expired`)),
                maxTime
            )
        );

        return Promise.race([loadTheme, timeout]);
    }

    /**
     * Loads the CSS Vars ponyfill in case the browser is Internet Explorer 11. It can also
     * forces to load in modern browsers in case the user passes `onlyLegacy` property as false
     */
    static startCssVarsPonyfill(ponyfillOptions = { onlyLegacy: true }) {
        return new Promise((resolve, reject) => {
            if (!isInternetExplorer11() && ponyfillOptions.onlyLegacy === true) {
                resolve();
                return;
            }

            cssVarsPonyfill({
                ...ponyfillOptions,
                // eslint-disable-next-line no-unused-vars
                onComplete(cssText, node, url) {
                    resolve();
                },
                // eslint-disable-next-line no-unused-vars
                onError(message, node, xhr, url) {
                    reject(message);
                }
            });
        });
    }

    static applyTheme(themeURL) {
        return new Promise((resolve, reject) => {
            if (themeURL === '') {
                resolve();
                return;
            }

            const styleElement = document.createElement('link');
            styleElement.type = 'text/css';
            styleElement.rel = 'stylesheet';
            styleElement.href = themeURL;
            styleElement.onload = () => resolve();
            styleElement.onerror = () => reject(new Error(`Failed to load the theme: ${themeURL}`));

            const head = document.getElementsByTagName('head')[0];
            head.appendChild(styleElement);
        });
    }
}

export default ThemeLoader;
