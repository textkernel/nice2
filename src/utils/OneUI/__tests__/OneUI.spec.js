import cssVarsPonyfill from 'css-vars-ponyfill';
import { OneUI } from '../OneUI';

jest.mock('css-vars-ponyfill');

describe('OneUI loader that starts the ponyfill and attach the theme to DOM', () => {
    it('should attach <link> element with theme URL to DOM tree', () => {
        const themeURL = 'http://theme-cdn.com/mytheme.css';
        document.body.innerHTML = '<head></head>';

        OneUI.applyTheme(themeURL);

        expect(document.getElementsByTagName('link')[0].href).toBe(themeURL);
    });

    it('should load the ponyfill if browser is IE11', () => {
        const ieUserAgent = 'Mozilla/5.0(Windows NT 6.3; Trident/7.0; rv:11.0) like Gecko';

        Object.defineProperty(window.navigator, 'userAgent', {
            value: ieUserAgent,
            configurable: true,
        });

        OneUI.applyCssVarsPonyfill();

        expect(cssVarsPonyfill).toBeCalledTimes(1);
    });

    it('should not load the ponyfill if browser is different than IE11', () => {
        const chromeUserAgent =
            'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.96 Safari/537.36';

        Object.defineProperty(window.navigator, 'userAgent', {
            value: chromeUserAgent,
            configurable: true,
        });

        OneUI.applyCssVarsPonyfill();

        expect(cssVarsPonyfill).toBeCalledTimes(0);
    });

    it('should apply theme and start ponyfill on init', () => {
        const themeURL = 'http://theme-cdn.com/mytheme.css';
        const ponyfillOptions = { onlyLegacy: false };

        const mockedApplyTheme = jest.spyOn(OneUI, 'applyTheme');
        const mockedApplyCssVarsPonyfill = jest.spyOn(OneUI, 'applyCssVarsPonyfill');

        mockedApplyTheme.mockImplementation(() => Promise.resolve());
        mockedApplyCssVarsPonyfill.mockImplementation(() => Promise.resolve());

        OneUI.init({ themeURL, ponyfillOptions });

        expect(mockedApplyTheme).toBeCalledWith(themeURL);
        expect(mockedApplyCssVarsPonyfill).toBeCalledWith(ponyfillOptions);

        mockedApplyTheme.mockRestore();
        mockedApplyCssVarsPonyfill.mockRestore();
    });

    it('should not attach link element in the DOM when themeURL is not provided', () => {
        const themeURL = null;

        const documentHead = '<meta charset="utf-8">s';
        document.head.innerHTML = documentHead;

        OneUI.applyTheme(themeURL);

        expect(document.head.innerHTML).toEqual(documentHead);
    });

    it('should call the onComplete callback function when ponyfill is loaded', () => {
        cssVarsPonyfill.mockImplementationOnce(ponyfillOptions =>
            setTimeout(ponyfillOptions.onComplete, 10)
        );

        const ieUserAgent = 'Mozilla/5.0(Windows NT 6.3; Trident/7.0; rv:11.0) like Gecko';

        Object.defineProperty(window.navigator, 'userAgent', {
            value: ieUserAgent,
            configurable: true,
        });

        const ponyfillOptions = {
            onComplete: jest.fn(),
        };

        return OneUI.applyCssVarsPonyfill(ponyfillOptions).then(() => {
            expect(ponyfillOptions.onComplete).toBeCalledTimes(1);
        });
    });

    it('should call the onError callback function when ponyfill isnt loaded', () => {
        cssVarsPonyfill.mockImplementationOnce(ponyfillOptions =>
            setTimeout(ponyfillOptions.onError, 10)
        );

        const ieUserAgent = 'Mozilla/5.0(Windows NT 6.3; Trident/7.0; rv:11.0) like Gecko';

        Object.defineProperty(window.navigator, 'userAgent', {
            value: ieUserAgent,
            configurable: true,
        });

        const ponyfillOptions = {
            onError: jest.fn(),
        };

        return OneUI.applyCssVarsPonyfill(ponyfillOptions).catch(() => {
            expect(ponyfillOptions.onError).toBeCalledTimes(1);
        });
    });
});
