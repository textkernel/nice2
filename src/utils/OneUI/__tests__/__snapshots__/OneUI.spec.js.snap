// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OneUI loader that starts the ponyfill and attach the theme to DOM should throw an error when timeout expired 1`] = `"Theme \\"\\" not loaded. Loading time expired"`;
